File: src/utils/benchmark.ts
import type { Position } from '../types'
import { isClosedLot } from './rowDerivation'
import { priceAt, type TickerHistory } from './chartHistory'

export type BenchmarkKey = 'none' | 'msci' | 'sp500'

/** URTH is the iShares MSCI World ETF; ^GSPC the S&P 500 index. Both are price series. */
export const BENCHMARKS: Record<Exclude<BenchmarkKey, 'none'>, { ticker: string; label: string }> = {
  msci: { ticker: 'URTH', label: 'MSCI World' },
  sp500: { ticker: '^GSPC', label: 'S&P 500' },
}

type ConvertAt = (amount: number, from: string, to: string, date: string) => number

/**
 * "What if every buy and sell had gone into the benchmark instead" — a cash-flow-matched
 * (Long–Nickels PME) return line in the display currency, directly comparable with the
 * portfolio's own price-P&L line. Each buy buys benchmark units for the same money on the
 * same day; each sell withdraws the same money. Units can go negative when a sale returns
 * more than the benchmark would have grown to — the standard PME behaviour.
 *
 * Mirrored by core/calc/Benchmark.kt against test-fixtures/benchmark/cases.json.
 */
export function benchmarkSeries(
  positions: Position[],
  dates: string[],
  bench: TickerHistory,
  benchCurrency: string,
  displayCurrency: string,
  convertAt: ConvertAt,
): number[] {
  if (bench.length === 0) return []
  interface Flow { date: string; units: number; invested: number }
  const flows: Flow[] = []
  positions.forEach((p) => {
    const buyBench = priceAt(bench, p.buyDate)
    if (buyBench) {
      const cost = p.buyPrice * p.quantity
      flows.push({
        date: p.buyDate,
        units: convertAt(cost, p.currency, benchCurrency, p.buyDate) / buyBench,
        invested: convertAt(cost, p.currency, displayCurrency, p.buyDate),
      })
    }
    if (isClosedLot(p)) {
      const sellBench = priceAt(bench, p.sellDate!)
      if (sellBench) {
        const proceeds = p.sellPrice! * p.quantity
        flows.push({
          date: p.sellDate!,
          units: -convertAt(proceeds, p.currency, benchCurrency, p.sellDate!) / sellBench,
          invested: -convertAt(proceeds, p.currency, displayCurrency, p.sellDate!),
        })
      }
    }
  })
  flows.sort((a, b) => a.date.localeCompare(b.date))

  let i = 0
  let units = 0
  let invested = 0
  return dates.map((date) => {
    while (i < flows.length && flows[i].date <= date) {
      units += flows[i].units
      invested += flows[i].invested
      i++
    }
    const price = priceAt(bench, date) ?? 0
    return convertAt(units * price, benchCurrency, displayCurrency, date) - invested
  })
}

File: android/core/calc/src/main/kotlin/com/stocktracker/core/calc/Benchmark.kt
package com.stocktracker.core.calc

import com.stocktracker.core.model.Position
import com.stocktracker.core.model.PriceHistory

enum class Benchmark(val ticker: String, val label: String) {
    MSCI_WORLD("URTH", "MSCI World"),
    SP500("^GSPC", "S&P 500"),
}

/**
 * "What if every buy and sell had gone into the benchmark instead" — a cash-flow-matched
 * (Long–Nickels PME) return line in the display currency, comparable with the portfolio's own
 * price-P&L line. Units may go negative when a sale returns more than the benchmark would
 * have grown to. Mirrors `benchmarkSeries` in src/utils/benchmark.ts.
 */
fun benchmarkSeries(
    positions: List<Position>,
    dates: List<String>,
    bench: PriceHistory,
    benchCurrency: String,
    displayCurrency: String,
    convertAt: (Double, String, String, String) -> Double,
): List<Double> {
    if (bench.isEmpty()) return emptyList()
    data class Flow(val date: String, val units: Double, val invested: Double)
    val flows = mutableListOf<Flow>()
    positions.forEach { p ->
        priceAt(bench, p.buyDate)?.takeIf { it != 0.0 }?.let { buyBench ->
            val cost = p.buyPrice * p.quantity
            flows += Flow(
                p.buyDate,
                convertAt(cost, p.currency, benchCurrency, p.buyDate) / buyBench,
                convertAt(cost, p.currency, displayCurrency, p.buyDate),
            )
        }
        if (isClosedLot(p)) {
            val sellDate = p.sellDate!!
            priceAt(bench, sellDate)?.takeIf { it != 0.0 }?.let { sellBench ->
                val proceeds = p.sellPrice!! * p.quantity
                flows += Flow(
                    sellDate,
                    -convertAt(proceeds, p.currency, benchCurrency, sellDate) / sellBench,
                    -convertAt(proceeds, p.currency, displayCurrency, sellDate),
                )
            }
        }
    }
    flows.sortBy { it.date }

    var i = 0
    var units = 0.0
    var invested = 0.0
    return dates.map { date ->
        while (i < flows.size && flows[i].date <= date) {
            units += flows[i].units
            invested += flows[i].invested
            i++
        }
        val price = priceAt(bench, date) ?: 0.0
        convertAt(units * price, benchCurrency, displayCurrency, date) - invested
    }
}
